### PR TITLE
[add pipeline]: A new Pixart-based inpainting pipeline.

### PR DESCRIPTION
**Pull request by eightmusic**
We add a pixart model-based inpaint method that works similarly to pipeline_stable_diffusion_inpaint. @sayakpaul @yiyixuxu @DN6

### CODE REVIEW

**Comment by yiyixuxu** [comment]
cc @asomoza here too!

**Comment by sayakpaul** [comment]
@lawrence-cj could you review this too?

File: src/diffusers/pipelines/pixart_alpha/pipeline_pixart_inpaint.py [hunk 1]
```diff
@@ -0,0 +1,1097 @@
...
+
+        >>> def download_image(url):
+        ...     response = requests.get(url)
+        ...     return PIL.Image.open(BytesIO(response.content)).convert("RGB")
+
+
+        >>> img_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo.png"
+        >>> mask_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo_mask.png"
+
+        >>> init_image = download_image(img_url).resize((512, 512))
```
Comment (sayakpaul) [inline src/diffusers/pipelines/pixart_alpha/pipeline_pixart_inpaint.py:67]: Could replace this with `diffusers.utils.load_image()`. Less lines of code. WDYT?

File: src/diffusers/pipelines/pixart_alpha/pipeline_pixart_inpaint.py [hunk 2]
```diff
@@ -0,0 +1,1097 @@
...
+        ...     "PixArt-alpha/PixArt-XL-2-1024-MS", torch_dtype=torch.float16
+        ... )
+        >>> pipe = pipe.to("cuda")
+
+        >>> prompt = "Face of a yellow cat, high resolution, sitting on a park bench"
+        >>> image = pipe(prompt=prompt, image=init_image, mask_image=mask_image).images[0]
+        ```
+"""
+
+ASPECT_RATIO_1024_BIN = {
```
Comment (sayakpaul) [inline src/diffusers/pipelines/pixart_alpha/pipeline_pixart_inpaint.py:80]: Do we wanna move these constants to the init of `pixart` as these are now access by two pipelines? WDYT?

Cc: @lawrence-cj WDYT?

File: src/diffusers/pipelines/pixart_alpha/pipeline_pixart_inpaint.py [hunk 3]
```diff
@@ -0,0 +1,1097 @@
...
+            negative_prompt: str = "",
+            num_images_per_prompt: int = 1,
+            device: Optional[torch.device] = None,
+            prompt_embeds: Optional[torch.FloatTensor] = None,
+            negative_prompt_embeds: Optional[torch.FloatTensor] = None,
+            prompt_attention_mask: Optional[torch.FloatTensor] = None,
+            negative_prompt_attention_mask: Optional[torch.FloatTensor] = None,
+            clean_caption: bool = False,
+            **kwargs,
+    ):
```
Comment (sayakpaul) [inline src/diffusers/pipelines/pixart_alpha/pipeline_pixart_inpaint.py:294]: We should use the same method from the PixArtAlpha implementation and use a `# Copied from ...` statement here.

File: src/diffusers/pipelines/pixart_alpha/pipeline_pixart_inpaint.py [hunk 4]
```diff
@@ -0,0 +1,1097 @@
...
+        if accepts_eta:
+            extra_step_kwargs["eta"] = eta
+
+        # check if the scheduler accepts generator
+        accepts_generator = "generator" in set(inspect.signature(self.scheduler.step).parameters.keys())
+        if accepts_generator:
+            extra_step_kwargs["generator"] = generator
+        return extra_step_kwargs
+
+    def check_inputs(
```
Comment (sayakpaul) [inline src/diffusers/pipelines/pixart_alpha/pipeline_pixart_inpaint.py:439]: If this copied from the text-to-image pipeline implementation, let's use the `# Copied from ...` statement here. Applies here and elsewhere. 

But here, I think the function should also validate the input image and the mask image. Here's a reference:

https://github.com/huggingface/diffusers/blob/6c60e430eeb050535aee8854c6ad649b162b62a9/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_inpaint.py#L778

File: src/diffusers/pipelines/pixart_alpha/pipeline_pixart_inpaint.py [hunk 5]
```diff
@@ -0,0 +1,1097 @@
...
+    @staticmethod
+    def classify_height_width_bin(height: int, width: int, ratios: dict) -> Tuple[int, int]:
+        """Returns binned height and width."""
+        ar = float(height / width)
+        closest_ratio = min(ratios.keys(), key=lambda ratio: abs(float(ratio) - ar))
+        default_hw = ratios[closest_ratio]
+        return int(default_hw[0]), int(default_hw[1])
+
+    @staticmethod
+    def resize_and_crop_tensor(samples: torch.Tensor, new_width: int, new_height: int) -> torch.Tensor:
```
Comment (sayakpaul) [inline src/diffusers/pipelines/pixart_alpha/pipeline_pixart_inpaint.py:701]: `# Copied from ...` statement missing.

**Review by sayakpaul — COMMENTED** [review]
The pipeline implementation looks very nice to me, thank you so much! 

I think what's pending for this PR to get merged is the following:

* Docs
* Test

File: src/diffusers/pipelines/pixart_alpha/pipeline_pixart_inpaint.py (hunk 5 above)
Comment (eightmusic) [reply src/diffusers/pipelines/pixart_alpha/pipeline_pixart_inpaint.py]: Recommits have been made in the appropriate format.@sayakpaul

File: src/diffusers/pipelines/pixart_alpha/pipeline_pixart_inpaint.py (hunk 2 above)
Comment (lawrence-cj) [reply src/diffusers/pipelines/pixart_alpha/pipeline_pixart_inpaint.py:73]: It sounds reasonable for me.

File: src/diffusers/pipelines/pixart_alpha/pipeline_pixart_inpaint.py [hunk 6]
```diff
@@ -0,0 +1,1092 @@
...
+            max_sequence_length (`int` defaults to 120): Maximum sequence length to use with the `prompt`.
+
+        Examples:
+
+        Returns:
+            [`~pipelines.ImagePipelineOutput`] or `tuple`:
+                If `return_dict` is `True`, [`~pipelines.ImagePipelineOutput`] is returned, otherwise a `tuple` is
+                returned where the first element is a list with the generated images
+        """
+        if "mask_feature" in kwargs:
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/pixart_alpha/pipeline_pixart_inpaint.py:886]: we don't need to deprecate for new pipeline

**Review by yiyixuxu — COMMENTED** [review]
thanks for adding this pipeline!
+1 on @sayakpaul 's comments about doc and tests

**Comment by sayakpaul** [comment]
@eightmusic let's get the comments resolved and we would be happy to include in the new release we're cooking.

**Comment by eightmusic** [comment]
> @eightmusic let's get the comments resolved and we would be happy to include in the new release we're cooking.
ruff check examples scripts src tests utils benchmarks setup.py
src/diffusers/pipelines/pixart_alpha/pipeline_pixart_inpaint.py:15:1: I001 [*] Import block is un-sorted or un-formatted
src/diffusers/pipelines/pixart_alpha/pipeline_pixart_inpaint.py:23:31: F401 [*] `torch.nn.functional` imported but unused
src/diffusers/pipelines/pixart_alpha/pipeline_pixart_inpaint.py:454:1: W293 [*] Blank line contains whitespace
src/diffusers/pipelines/pixart_alpha/pipeline_pixart_inpaint.py:509:1: W293 [*] Blank line contains whitespace
src/diffusers/pipelines/pixart_alpha/pipeline_pixart_inpaint.py:1092:49: W292 [*] No newline at end of file
Found 5 errors.
[*] 5 fixable with the `--fix` option.
make: *** [Makefile:43: quality] Error 1
These problems have been solved.@sayakpaul

**Comment by sayakpaul** [comment]
There are still open comments left to be addressed. Additionally, we need to have docs and tests for this to get merged.

**Comment by eightmusic** [comment]
> Thank you!
> 
> There are still open comments left to be addressed. Additionally, we need to have docs and tests for this to get merged.

The previous comments should be resolved, is there anything else I need to do?

**Comment by sayakpaul** [comment]
Well, the tests are failing. We need to get them sorted. And then there are comments still open, e.g.: https://github.com/huggingface/diffusers/pull/7929#pullrequestreview-2066566910, https://github.com/huggingface/diffusers/pull/7929#discussion_r1604487436. 

And then we need to add tests for the pipeline and documentation. LMK if there's anything unclear.

**Comment by eightmusic** [comment]
https://github.com/huggingface/diffusers/pull/7929#pullrequestreview-2066566910, https://github.com/huggingface/diffusers/pull/7929#discussion_r1604487436.This problem has just been solved.'We need to get them sorted.' What do I need to do here. Are there any references to test and docs? I don't know what this is.@sayakpaul

**Comment by sayakpaul** [comment]
> Are there any references to test and docs? 

* Here's how an inpainting test suite should be devised: https://github.com/huggingface/diffusers/blob/main/tests/pipelines/stable_diffusion_2/test_stable_diffusion_inpaint.py. 
* Here's how a pipeline doc should be added: https://github.com/huggingface/diffusers/blob/main/docs/source/en/api/pipelines/pixart.md. 


Quality tests can be fixed by running `make style && make quality`. Refer to the [contribution guidelines](https://huggingface.co/docs/diffusers/en/conceptual/contribution) for more details.

**Comment by lawrence-cj** [comment]
Sry to disturb you, but I found I can't push a commit to this branch, any suggestions? @sayakpaul

**Comment by sayakpaul** [comment]
I think you will need for submit a PR to https://github.com/eightmusic/diffusers/. Perhaps @eightmusic could add you as a collaborator to the repo?

**Comment by lawrence-cj** [comment]
Agree. I can help adding the make style and test part. Is it possible to add me as a collaborator to your forked diffusers repo?

**Comment by eightmusic** [comment]
> Agree. I can help adding the make style and test part. Is it possible to add me as a collaborator to your forked diffusers repo? @eightmusic

of course.
